File: backend/app/schemas/recommendation_page.py
from typing import Optional
from uuid import UUID
from pydantic import BaseModel


# ── Formula accord item ──────────────────────────────────────────
class AccordFormulaItem(BaseModel):
    accord_id: str
    accord_name: str
    volume_ml: float


# ── Single combination shown on recommendation page ──────────────
class RecommendationCombination(BaseModel):
    combo_id: str               # e.g. "comb_1" — from LLM output id field
    recommendation_rank: int
    is_default_selected: bool
    status: str                 # not_selected | selected | removed | purchased
    combo_name: str
    description: str
    justification: str
    formula: list[AccordFormulaItem]
    quantity: int
    total_volume_ml: float


# ── Response for GET /recommendation/page-data ───────────────────
class RecommendationPageResponse(BaseModel):
    success: bool
    session_id: Optional[str] = None
    combinations: Optional[list[RecommendationCombination]] = None
    message: Optional[str] = None


# ── Request for POST /recommendation/confirm ─────────────────────
class ConfirmSelectionRequest(BaseModel):
    session_id: UUID
    # combo_ids of the combinations the user selected (e.g. ["comb_1", "comb_3"])
    selected_combo_ids: list[str]
    # Optional qty per combo_id: { "comb_1": 2, "comb_2": 1 }
    qty_updates: Optional[dict[str, int]] = None


# ── Response for POST /recommendation/confirm ────────────────────
class ConfirmSelectionResponse(BaseModel):
    success: bool
    message: str
    selected_count: Optional[int] = None

File: frontend/src/services/api/recommendation-page.ts
const API_BASE_URL = "http://localhost:8000";

// ── Types ─────────────────────────────────────────────────────────

export interface AccordFormulaItem {
  accord_id: string;
  accord_name: string;
  volume_ml: number;
}

export interface RecommendationCombination {
  combo_id: string;            // e.g. "comb_1" — identifier for selection
  recommendation_rank: number;
  is_default_selected: boolean;
  status: string;
  combo_name: string;
  description: string;
  justification: string;
  formula: AccordFormulaItem[];
  quantity: number;
  total_volume_ml: number;
}

export interface RecommendationPageData {
  success: boolean;
  session_id: string;
  combinations: RecommendationCombination[];
  message?: string;
}

export interface ConfirmSelectionPayload {
  session_id: string;
  selected_combo_ids: string[];            // e.g. ["comb_1", "comb_3"]
  qty_updates?: Record<string, number>;   // e.g. { "comb_1": 2 }
}

export interface ConfirmSelectionResponse {
  success: boolean;
  message: string;
  selected_count?: number;
}

// ── Fetch recommendation page data ───────────────────────────────

export async function fetchRecommendationPageData(
  sessionId: string
): Promise<RecommendationPageData> {
  const token = localStorage.getItem("jwt_token");

  const response = await fetch(
    `${API_BASE_URL}/recommendation/page-data?session_id=${sessionId}`,
    {
      method: "GET",
      headers: {
        Authorization: `Bearer ${token}`,
      },
    }
  );

  if (!response.ok) {
    const errorData = await response.json();
    console.log("RECOMMENDATION PAGE DATA ERROR", errorData);
    throw new Error(errorData?.detail || "Failed to load recommendations.");
  }

  return response.json();
}

// ── Confirm cart selection ────────────────────────────────────────

export async function confirmSelection(
  payload: ConfirmSelectionPayload
): Promise<ConfirmSelectionResponse> {
  const token = localStorage.getItem("jwt_token");

  const response = await fetch(`${API_BASE_URL}/recommendation/confirm`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const errorData = await response.json();
    console.log("CONFIRM SELECTION ERROR", errorData);
    throw new Error(errorData?.detail || "Failed to confirm selection.");
  }

  return response.json();
}
